File: internal/bot/random.go
package bot

import (
	"context"
	"errors"
	"math/rand"
	"time"

	"github.com/Mgrdich/TermChess/internal/engine"
)

// randomEngine implements the Easy bot using random move selection.
type randomEngine struct {
	name      string
	timeLimit time.Duration
	closed    bool
}

// SelectMove returns a move using weighted selection (70% tactical bias).
func (e *randomEngine) SelectMove(ctx context.Context, board *engine.Board) (engine.Move, error) {
	if e.closed {
		return engine.Move{}, errors.New("engine is closed")
	}

	// Get all legal moves
	moves := board.LegalMoves()
	if len(moves) == 0 {
		return engine.Move{}, errors.New("no legal moves available")
	}

	// If only one move, return it immediately (forced move)
	if len(moves) == 1 {
		return moves[0], nil
	}

	// Create timeout context
	ctx, cancel := context.WithTimeout(ctx, e.timeLimit)
	defer cancel()

	// Categorize moves
	captures := filterCaptures(board, moves)
	checks := filterChecks(board, moves)

	// Weighted selection: 70% tactical bias
	// Check context to respect timeout
	select {
	case <-ctx.Done():
		return engine.Move{}, ctx.Err()
	default:
		// 70% chance to pick a capture if available
		if rand.Float64() < 0.7 && len(captures) > 0 {
			return captures[rand.Intn(len(captures))], nil
		}

		// 50% chance to pick a check if available
		if rand.Float64() < 0.5 && len(checks) > 0 {
			return checks[rand.Intn(len(checks))], nil
		}

		// Fallback: any random legal move
		return moves[rand.Intn(len(moves))], nil
	}
}

// filterCaptures returns all moves that capture an opponent's piece.
func filterCaptures(board *engine.Board, moves []engine.Move) []engine.Move {
	var captures []engine.Move
	for _, m := range moves {
		// Check if destination square has an opponent piece
		targetPiece := board.PieceAt(m.To)
		if !targetPiece.IsEmpty() {
			captures = append(captures, m)
		}
	}
	return captures
}

// filterChecks returns all moves that give check to the opponent's king.
func filterChecks(board *engine.Board, moves []engine.Move) []engine.Move {
	var checks []engine.Move
	for _, m := range moves {
		// Make move on a copy and check if opponent is in check
		boardCopy := board.Copy()
		boardCopy.MakeMove(m)

		// After making the move, check if the NEW active color (opponent) is in check
		if boardCopy.InCheck() {
			checks = append(checks, m)
		}
	}
	return checks
}

// Name returns the human-readable name of this engine.
func (e *randomEngine) Name() string {
	return e.name
}

// Close releases resources held by the engine.
func (e *randomEngine) Close() error {
	e.closed = true
	return nil
}

// Info returns metadata about this engine.
func (e *randomEngine) Info() Info {
	return Info{
		Name:       e.name,
		Author:     "TermChess",
		Version:    "1.0",
		Type:       TypeInternal,
		Difficulty: Easy,
		Features: map[string]bool{
			"random_selection":   true,
			"tactical_awareness": true,
			"weighted_selection": true,
		},
	}
}

File: rust/crates/bot/src/random.rs
//! The Easy bot: weighted random move selection (ported from `random.go`).

use std::sync::atomic::{AtomicBool, Ordering};
use std::time::{Duration, Instant};

use engine::{Board, Move, PieceType, Square};
use rand::Rng;

use crate::context::Context;
use crate::error::EngineError;
use crate::interfaces::{Difficulty, Engine, EngineType, Info, Inspectable};

/// The Easy bot: picks moves with a 70% tactical bias toward captures and a
/// secondary 50% bias toward checks.
#[derive(Debug)]
pub struct RandomEngine {
    pub(crate) name: String,
    pub(crate) time_limit: Duration,
    closed: AtomicBool,
}

impl RandomEngine {
    /// Creates a new random engine with the given name and time limit.
    pub(crate) fn new(name: impl Into<String>, time_limit: Duration) -> RandomEngine {
        RandomEngine {
            name: name.into(),
            time_limit,
            closed: AtomicBool::new(false),
        }
    }
}

impl Engine for RandomEngine {
    fn select_move(&self, ctx: &Context, board: &Board) -> Result<Move, EngineError> {
        if self.closed.load(Ordering::SeqCst) {
            return Err(EngineError::Closed);
        }

        let moves = board.legal_moves();
        if moves.is_empty() {
            return Err(EngineError::NoLegalMovesAvailable);
        }

        // Forced move: return immediately.
        if moves.len() == 1 {
            return Ok(moves[0]);
        }

        // Derived deadline from the engine's own time limit, combined with the
        // caller's context (mirrors context.WithTimeout).
        let deadline = Instant::now() + self.time_limit;

        let captures = filter_captures(board, &moves);
        let checks = filter_checks(board, &moves);

        // Respect cancellation / timeout before selecting.
        if ctx.is_done() {
            return Err(ctx.err().unwrap_or(EngineError::ContextCanceled));
        }
        if Instant::now() >= deadline {
            return Err(EngineError::ContextDeadlineExceeded);
        }

        let mut rng = rand::thread_rng();

        // 70% chance to pick a capture if available.
        if rng.gen::<f64>() < 0.7 && !captures.is_empty() {
            return Ok(captures[rng.gen_range(0..captures.len())]);
        }

        // 50% chance to pick a check if available.
        if rng.gen::<f64>() < 0.5 && !checks.is_empty() {
            return Ok(checks[rng.gen_range(0..checks.len())]);
        }

        // Fallback: any random legal move.
        Ok(moves[rng.gen_range(0..moves.len())])
    }

    fn name(&self) -> &str {
        &self.name
    }

    fn close(&self) -> Result<(), EngineError> {
        self.closed.store(true, Ordering::SeqCst);
        Ok(())
    }
}

impl Inspectable for RandomEngine {
    fn info(&self) -> Info {
        let mut features = std::collections::HashMap::new();
        features.insert("random_selection".to_string(), true);
        features.insert("tactical_awareness".to_string(), true);
        features.insert("weighted_selection".to_string(), true);
        Info {
            name: self.name.clone(),
            author: "TermChess".to_string(),
            version: "1.0".to_string(),
            engine_type: EngineType::Internal,
            difficulty: Difficulty::Easy,
            features,
        }
    }
}

/// Returns all moves that capture an opponent's piece, including en passant.
pub(crate) fn filter_captures(board: &Board, moves: &[Move]) -> Vec<Move> {
    let mut captures = Vec::new();
    for &m in moves {
        // Normal capture: destination has an opponent piece.
        let target_piece = board.piece_at(m.to);
        if !target_piece.is_empty() {
            captures.push(m);
            continue;
        }

        // En passant capture: pawn moves to the en passant target square.
        let moving_piece = board.piece_at(m.from);
        if moving_piece.piece_type() == PieceType::Pawn
            && board.en_passant_sq >= 0
            && m.to == Square(board.en_passant_sq)
        {
            captures.push(m);
        }
    }
    captures
}

/// Returns all moves that give check to the opponent's king.
pub(crate) fn filter_checks(board: &Board, moves: &[Move]) -> Vec<Move> {
    let mut checks = Vec::new();
    for &m in moves {
        let mut board_copy = board.copy();
        let _ = board_copy.make_move(m);

        // After the move, is the new active color (opponent) in check?
        if board_copy.in_check() {
            checks.push(m);
        }
    }
    checks
}
